runtime-rs: built-in Dragonball sandbox part I - resource and device managers

studychao opened the pull request:
This pull request initializes the basic framework for Dragonball Sandbox and introduces the resource and device managers part.

In this PR, we introduce the abstractions for several management for virtual machine managers such as LegacyDeviceManager, AddressSpaceManager, ResourceManager and etc. They are responsible for maintaining different kinds of resources for virtual machines.

This is the first pull request for Dragonball Sandbox and there are approximately 6 PRs covering different parts of VMM in total. 

Fixes: #4257

wllenyj commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+
+/// Console manager to manage frontend and backend console devices.
+pub struct ConsoleManager {
+    pub(crate) epoll_mgr: EpollManager,
```
Can be done without pub(crate)

studychao replied:
Fixed.

quanweiZhou commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+        }
+
+        if should_drop {
+            let events = Events::with_data_raw(libc::STDIN_FILENO, EPOLL_EVENT_STDIN, EventSet::IN);
```
Is it need to set  `self.stdin_handle` to `None`?

studychao replied:
If self.stdin_handle is set to None, stdin related connection will be lost. We only want to drop the data in stdin channel here.

quanweiZhou reviewed (commented):
A few comments

wllenyj commented on src/dragonball/Makefile:
```diff
@@ -0,0 +1,27 @@
+
+test:
+	@echo "INFO: testing dragonball for development build"
+	cargo test --all-features -- --nocapture
```
need `--test-threads 1`
Otherwise, the kvm-based test case will fail.

studychao replied:
Changed.

studychao commented on src/dragonball/Cargo.toml:
```diff
@@ -0,0 +1,53 @@
+atomic-guest-memory = []
+virtio-vsock = ["dbs-virtio-devices/virtio-vsock", "virtio-queue"]
+
+[patch.'crates-io']
```
This will be removed at the final version of Dragonball:)

studychao commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,681 @@
+    InvalidResourceRange,
+
+    /// No source available.
+    #[error("no resource avialable")]
```
fixed

studychao commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+            stdin_handle,
+            sock_listener,
+            sock_conn: None,
+            logger: logger.new(slog::o!()),
```
Added.

studychao commented on src/dragonball/src/device_manager/mod.rs:
```diff
@@ -0,0 +1,603 @@
+
+/// Type of the dragonball virtio mmio devices.
+#[cfg(feature = "dbs-virtio-devices")]
+pub type DbsMmioV2Device = dbs_virtio_devices::mmio::MmioV2Device<
```
Changed. Thanks.

liubin commented on src/dragonball/README.md:
```diff
@@ -0,0 +1,34 @@
+which is optimized for container workloads with:
+- container image management and acceleration service
+- flexible and high-performance virtual device drivers
+- low cpu & memory overhead
```
cpu -> CPU

studychao replied:
Thanks, fixed:)

liubin commented on README.md:
```diff
@@ -118,6 +118,7 @@ The table below lists the core parts of the project:
 | [runtime](src/runtime) | core | Main component run by a container manager and providing a containerd shimv2 runtime implementation. |
 | [agent](src/agent) | core | Management process running inside the virtual machine / POD that sets up the container environment. |
 | [libraries](src/libs) | core | Library crates shared by multiple Kata Container components or published to [`crates.io`](https://crates.io/index.html) |
+| [dragonball](src/dragonball) | core | An optional built-in VMM brings out-of-the-box Kata containers experience with optimizations on container workloads |
```
containers -> Containers

studychao replied:
fixed.

liubin commented on src/dragonball/README.md:
```diff
@@ -0,0 +1,34 @@
+
+## Supported Architectures
+- x86-64
+- AArch64
```
AArch64 -> aarch64

studychao replied:
Thanks, fixed.

jodh-intel commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+
+use super::{DeviceMgrError, Result};
+
+const EPOLL_EVENT_SERIAL: u32 = 0;
+const EPOLL_EVENT_SERIAL_DATA: u32 = 1;
+const EPOLL_EVENT_STDIN: u32 = 2;
```
Could these be an `enum` instead?

studychao replied:
Since we have 
```
 match slot {
            EPOLL_EVENT_SERIAL => {
                if let Err(e) = self.uds_listener_accept(ops) {
                    slog::warn!(self.logger, "failed to accept incoming connection, {:?}", e);
                }
            }
            EPOLL_EVENT_SERIAL_DATA => {
                if let Err(e) = self.uds_read_in(ops) {
                    slog::warn!(self.logger, "failed to read data from UDS, {:?}", e);
                }
            }
            EPOLL_EVENT_STDIN => {
                if let Err(e) = self.stdio_read_in(ops) {
                    slog::warn!(self.logger, "failed to read data from stdin, {:?}", e);
                }
            }
```
it will be kinda complex if we match slot(u32) with the enum

jodh-intel commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+    epoll_mgr: EpollManager,
+    logger: slog::Logger,
+    subscriber_id: Option<SubscriberId>,
+    stdin_handle: Option<std::io::Stdin>,
+    sock_path: Option<String>,
```
It appears (?) that a console manager must either set the `stdin_handle` or `sock_path`, but not both. If true, This could be an enum containing either the `std::io::Stdin` or the socket path `String`. That would make the code clearer I think.

studychao replied:
Yes, I have changed that to an enum.

jodh-intel commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+    fn stdio_read_in(&mut self, ops: &mut EventOps) -> std::io::Result<()> {
+        let mut should_drop = true;
+
+        if self.stdin_handle.is_some() {
+            let mut out = [0u8; 64];
+            // Safe to unwrap() because self.stdin_handle is Some().
+            let stdin_lock = self.stdin_handle.as_ref().unwrap().lock();
```
This would be better coded using `let ... if` to avoid the check-and-unwrap combination:

```rust
let handle = if let Some(self.stdin_handle) {
  // ...
}
```

Same comment for other `unwrap()` calls.

studychao replied:
Changed.

jodh-intel commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,427 @@
+        let mut should_drop = true;
+
+        if let Some(conn_sock) = self.sock_conn.as_mut() {
+            let mut out = [0u8; 64];
```
Could you make this a `const` with a comment explaining why 64 was chosen?

studychao replied:
Changed.

jodh-intel commented on src/dragonball/src/error.rs:
```diff
@@ -0,0 +1,41 @@
+// Copyright (C) 2022 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
```
Please can you provide details in the commit(s) showing where this code originated (project, URL, version / release, date). For example, was the Amazon copyright material taken from Firecracker, or some other project?

Same comment for the other files with 3rd party copyright statements.

studychao replied:
`Acknowledgement` part added in the README in dragonball helps explain this :)

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+
+// The LEGACY_IRQ_BASE irq is reserved for shared IRQ.
+const SHARED_IRQ: u32 = LEGACY_IRQ_BASE;
+#[cfg(target_arch = "x86_64")]
+const MSI_IRQ_BASE: u32 = 24;
+#[cfg(target_arch = "aarch64")]
```
To avoid any confusion, can you add a `cfg` check so that if the target arch is not one of these two, the build fails with a message?

Could you also reference a GitHub issue (raise one if necessary) that lists the Kata architectures this code does not yet run on, along with some details of what would need to be done for those platforms to get dragonball working.

studychao replied:
Sure, we'll do that check in another PR. Issue #4316 is created for this.

Besides, the unsupported platforms issue is also created. #4320

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+/// It's gsi base id of msi
+const MSI_IRQ_BASE: u32 = LEGACY_IRQ_MAX + 1;
+
+// Currently, The max number of MSI IRQ we support is 1023
```
Could you explain this limit in the comment?

studychao replied:
Added.

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,681 @@
+// Linux: KVM_MAX_IRQ_ROUTES is defined as 4096 in linux/include/linux/kvm_host.h
+const MSI_IRQ_MAX: u32 = 1023;
+// Linux: KVM_USER_MEM_SLOTS is 509 for x86_64 and 512 for aarch64.
+const KVM_USER_MEM_SLOTS: u32 = 509;
```
Where does this value come from? Could you add a reference to the kernel source line that specifies this value?

jongwu replied:
Yeah, I'm curious about it too. I check it in aarch64 and it's 32767. If I'm wrong, pls correct me.

xuejun-xj replied:
Thans for your comments!
As Dragonball only supports kernel-4.19 and kernel-5.10 currently, the macro "KVM_USER_MEM_SLOTS" is defined in file linux/arch/arm64/include/asm/kvm_host.h, in which KVM_USER_MEM_SLOTS = 512. 
In kernel-5.12 and above, this macro is defined in linux/include/linux/kvm_host.h and KVM_USER_MEM_SLOTS = 32767.

jongwu replied:
Thanks for explanation.
Does that mean we must run kata3.0 on top of specific host kernel version?

xuejun-xj replied:
@jongwu :)
For now, we only test Arm Dragonball on these two kernel version(4.19 & 5.10). In the future, we may support higher versions of kernel.

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+
+impl Default for ResoureManagerBuilder {
+    fn default() -> Self {
+        ResoureManagerBuilder {
+            legacy_irq_pool: IntervalTree::new(),
+            msi_irq_pool: IntervalTree::new(),
+            pio_pool: IntervalTree::new(),
+            mmio_pool: IntervalTree::new(),
+            mem_pool: IntervalTree::new(),
+            kvm_mem_slot_pool: IntervalTree::new(),
+        }
```
`IntervalTree` provides `Default`, so this could become `ResoureManagerBuilder::default()` if you add `#[Default]` to `ResoureManagerBuilder` I think?

studychao replied:
changed

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+    NoAvailResource,
+}
+
+struct ResoureManagerBuilder {
+    legacy_irq_pool: IntervalTree<()>,
+    msi_irq_pool: IntervalTree<()>,
+    pio_pool: IntervalTree<()>,
+    mmio_pool: IntervalTree<()>,
+    mem_pool: IntervalTree<()>,
+    kvm_mem_slot_pool: IntervalTree<()>,
+}
```
Please document all the element.

studychao replied:
Added.

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+                // devices, such as LAPIC, IOAPIC, HPET etc, in this range. And we don't explicitly
+                // allocate MMIO address for those devices.
+                // Reserve 16KB more for passthrough ipi.
+                let constraint = Constraint::new(0x400_4000u64)
+                    .min(0x1_0000_0000u64 - 0x400_4000u64)
+                    .max(0xffff_ffffu64);
```
Could you define some `const`'s to help explain these values and avoid repetition?

studychao replied:
Added.

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+        }
+
+        if !(LEGACY_IRQ_BASE..=LEGACY_IRQ_MAX).contains(&irq) {
+            panic!("invalid irq number when freeing legacy irq");
```
If we really have to panic, please add a unit test using `#[should_panic]` for this scenario.

studychao replied:
The panic has been removed.

We use `Result` to return errors.

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+            || irq.checked_add(count).is_none()
+            || irq + count - 1 > MSI_IRQ_MAX
+        {
+            panic!("invalid irq number when freeing legacy irq");
```
Again, please add a unit test using `#[should_panic] ` for this and any other functions that panic.

studychao replied:
The panic has been removed.

jodh-intel commented on src/dragonball/src/resource_manager.rs:
```diff
@@ -0,0 +1,683 @@
+    pub fn free_device_resources(&self, resources: &DeviceResources) {
+        for res in resources.iter() {
+            match res {
+                Resource::PioAddressRange { base, size } => self.free_pio_address(*base, *size),
+                Resource::MmioAddressRange { base, size } => self.free_mmio_address(*base, *size),
+                Resource::MemAddressRange { base, size } => self.free_mem_address(*base, *size),
+                Resource::LegacyIrq(base) => self.free_legacy_irq(*base),
+                Resource::MsiIrq { ty: _, base, size } => self.free_msi_irq(*base, *size),
+                Resource::KvmMemSlot(slot) => self.free_kvm_mem_slot(*slot),
+                Resource::MacAddresss(_) => {}
+            }
+        }
+    }
+
+    fn free_allocated_resources(
+        &self,
+        resources: DeviceResources,
+    ) -> Result<DeviceResources, ResourceError> {
+        self.free_device_resources(&resources);
+        Err(ResourceError::NoAvailResource)
```
This is slightly odd code: All the `free_*()` handlers `panic` on error and the function that calls the handlers then returns an error "on success". Also, if the resources are freed, saying there are no resources available is "correct", but it's potentially misleading as it really means, "there (probably) _are_ system resources to _create_ the requested resources, but those requested resources haven't been (re-)created yet".

Why can we just use the conventional methodology here and make the handlers return a `Result` to allow `free_device_resources()` to return that `Result`. This function could then return that `Result` *on error*, or `Ok` on success, as is normal.

studychao replied:
Thanks for your suggestion.

We have refactored this part.

jodh-intel commented on src/dragonball/src/vm/kernel_config.rs:
```diff
@@ -0,0 +1,82 @@
+pub struct KernelConfigInfo {
+    /// The descriptor to the kernel file.
+    kernel_file: File,
+    /// The descriptor to the initrd file, if there is one
+    initrd_file: Option<File>,
```
Does this imply that normal rootfs images cannot be used? Also, what does it mean if `initrd_file == None`?

studychao replied:
We support normal rootfs images and initrd is for providing the possibility for any case if users need to use initrd

studychao replied:
rootfs image related setting will be introduced in the next PR related to block device manager

jodh-intel commented on src/dragonball/src/vm/kernel_config.rs:
```diff
@@ -0,0 +1,82 @@
+    initrd_file: Option<File>,
+    /// The commandline for guest kernel.
+    cmdline: linux_loader::cmdline::Cmdline,
+    /// Share guest kernel text/ro sections.
+    share_ro_sections: String,
```
Why is this a `String` and not just a `bool`?

studychao replied:
This is a performance feature and we decide to not include this in the stage one for Dragonball:)

jodh-intel commented on src/dragonball/src/vm/mod.rs:
```diff
@@ -0,0 +1,20 @@
+    /// memory size for this region
+    pub size: u64,
+    /// numa node id on host for this region
+    pub host_numa_node_id: Option<u32>,
+    /// numa node id on guest for this region
+    pub guest_numa_node_id: Option<u32>,
```
- Are these values always different?
- Can multiple `NumaRegionInfo` object contain the same values for these elements?

studychao replied:
- host_numa_node_id and guest_numa_node_id could be the same or different, it's related to users' actual needs.
- Technically, it won't be multiple `NumaRegionInfo` that contain the same value. We will introduce the validation check in vm configuration part PR.

jodh-intel reviewed (commented):
Thanks @studychao - A few initial comments...

studychao commented:
Hi @jodh-intel thanks a lot for your comments. Your advice are very important to us.
I'll address your comments soon. :)

wllenyj commented on src/dragonball/Makefile:
```diff
@@ -0,0 +1,28 @@
+default: build
+
+build:
+    cargo update -p vm-memory:0.8.0 --precise 0.7.0
```
This is because the crates in dbs-xxx uses version 0.7.0 of vm-memory, but rust-vmm uses 0.8.0. This will be resolved by upgrading dbs-xxx in a later patch.

studychao replied:
space is changed to tab

wllenyj commented on src/dragonball/Makefile:
```diff
@@ -0,0 +1,28 @@
+
+test:
+	@echo "INFO: testing dragonball for development build"
+	cargo test --all-features -- --test-threads 1 --nocapture
```
This problem has been solved, and the `--test-threads 1` will be deleted in the following patch.

studychao replied:
Already deleted.

jiangliu commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+            )
+        };
+        if res < 0 {
+            warn!("failed to mbind memory to host_numa_node_id {}", node_id);
+        }
+        Ok(())
```
We treat mbind as a hint for performance improvement, but don't want to fail the pod:)

studychao replied:
Changed.

studychao commented on src/dragonball/README.md:
```diff
@@ -0,0 +1,32 @@
+# Introduction
+`Dragonball Sandbox` is a light-weight virtual machine manager(VMM) based on Linux Kernel-based Virtual Machine(KVM),
```
Changed:) Thanks.

studychao commented on src/dragonball/README.md:
```diff
@@ -0,0 +1,32 @@
+which is optimized for container workloads with:
+- container image management and acceleration service
+- flexible and high-performance virtual device drivers
+- low CPU & memory overhead
```
Changed

studychao commented on src/dragonball/README.md:
```diff
@@ -0,0 +1,32 @@
+and configuration process.
+
+# Getting Start
+[TODO]
```
Github issue added #4302

studychao commented on src/dragonball/docs/device.md:
```diff
@@ -0,0 +1,15 @@
+Currently we have following device manager:
+1. address space manager: abstracts virtual machine's physical management and provide mapping for guest virtual memory and MMIO ranges of emulated virtual devices, pass-through devices and vCPU.
+2. config manager: provides abstractions for configuration information.
+3. console manager: provides management for all console devices
```
Added:)

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+
+impl<'a> AddressSpaceMgrBuilder<'a> {
+    /// Create a new [`CreateParam`] object.
+    pub fn new(mem_type: &'a str, mem_file: &'a str) -> Self {
```
mem_file check will be introduced later in the following PRs about vm configurations.

mem_type check is added.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+                reg.start_addr().raw_value(),
+                host_addr
+            );
+            // Safe because the guest regions are guaranteed not to overlap.
```
The Interval Tree data structure provided by the `dbs-allocator` could guarantee the guest memory regions that it allocates won't overlay.

studychao replied:
unit test `test_allocate_duplicate_memory` added at resource_manager.rs to ensure that the memory regions sent to `map_to_kvm` cannot overlap.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+    }
+
+    /// Mmap the address space region into current process.
+    #[allow(clippy::type_complexity)]
```
Thanks:) We have simplify the type below so this clippy allow is no longer needed.:)

studychao commented on src/dragonball/README.md:
```diff
@@ -0,0 +1,32 @@
+- container image management and acceleration service
+- flexible and high-performance virtual device drivers
+- low CPU & memory overhead
+- industry-leading startup time
```
Changed~

studychao commented on src/dragonball/docs/device.md:
```diff
@@ -0,0 +1,15 @@
+1. address space manager: abstracts virtual machine's physical management and provide mapping for guest virtual memory and MMIO ranges of emulated virtual devices, pass-through devices and vCPU.
+2. config manager: provides abstractions for configuration information.
+3. console manager: provides management for all console devices
+4. resource manager: provides resource management for `legacy_irq_pool`, `msi_irq_pool`, `pio_pool`, `mmio_pool`, `mem_pool`, `kvm_mem_slot_pool`
```
Change this part to a table

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+const PRE_ALLOC_GRANULARITY: u64 = 32;
+
+// Maximal NUMA node ids on host
+const MAX_NODE: u32 = 64;
```
Explanation added

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+    SetFileSize(#[source] std::io::Error),
+
+    /// Failed to unlink memory file.
+    #[error("cannot unlike memory file")]
```
fixed~ thanks~

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+
+    /// Failure in initializing guest memory.
+    #[error("failure while initializing guest memory")]
+    MemoryNotInitialized,
```
Changed and moved.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+    }
+
+    /// Enable/disable adding numbered suffix to memory file path.
+    pub fn set_file_suffix(&mut self, enabled: bool) {
```
changed.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+
+    /// Failed to access address_space.
+    #[error("failed to access address_space")]
+    AddressSpaceNotInitialized,
```
- Actually, there is no difference, so I merge those two errors into one
- The comments could not be deleted since it's necessary to comment every variant for the pub enum or else rust will have warning about that :)

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+    }
+
+    /// Enable/disable memory pre-allocation.
+    pub fn set_prealloc(&mut self, prealloc: bool) {
```
changed. Thanks~

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+    }
+
+    /// Enable/disable KVM dirty page logging.
+    pub fn set_dirty_page_logging(&mut self, logging: bool) {
```
Changed~

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+
+    /// Set KVM [`VmFd`] handle to configure memory slots.
+    pub fn set_kvm_vm_fd(&mut self, vmfd: Arc<VmFd>) {
+        self.vmfd = Some(vmfd);
```
existing value return added.

studychao commented on src/dragonball/src/vm/mod.rs:
```diff
@@ -0,0 +1,20 @@
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
+pub struct NumaRegionInfo {
+    /// memory size for this region
+    pub size: u64,
```
Added:)

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+            let size = info.size << 20;
+
+            // Guest memory does not intersect with the MMIO hole.
+            // TODO: make it work for ARM
```
Added

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+                // Add guest memory below the MMIO hole, avoid splitting the memory region
+                // if the available address region is too small.
+                let mut below_size = dbs_boot::layout::MMIO_LOW_START - start_addr;
+                if below_size < (128 << 20) {
```
Added

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+    fn create_region(
+        &mut self,
+        start_addr: u64,
+        size: u64,
```
Added

studychao replied:
Thanks, changed.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+        // Touch every 4k page to trigger allocation. The step is 4K instead of 2M to ensure
+        // pre-allocation when running out of huge pages.
+        if region.perm_flags() & libc::MAP_POPULATE > 0 {
+            const PAGE_SIZE: u64 = 4096;
```
Because the memory pre-allocation process only wants to touch every 4K page instead of 2M or other page size settings to avoid running out of huge pages so we fix that const to 4096

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+        // pre-allocation when running out of huge pages.
+        if region.perm_flags() & libc::MAP_POPULATE > 0 {
+            const PAGE_SIZE: u64 = 4096;
+            const PAGE_SHIFT: u32 = 12;
```
Comment is added for explaining `PAGE_SHIFT`

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+        Ok(())
+    }
+
+    fn configure_thp_and_prealloc(
```
Added

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,829 @@
+            let addr = mmap_reg.as_ptr() as u64;
+            let npage = (mmap_reg.size() as u64) >> PAGE_SHIFT;
+
+            let mut touch_thread = ((mmap_reg.size() as u64) >> PRE_ALLOC_GRANULARITY) + 1;
```
This is for controlling the actual number of pre-allocating threads. After several performance tests, we decide to use one thread to do pre-allocating for every 4G memory.

studychao commented on src/dragonball/src/config_manager.rs:
```diff
@@ -0,0 +1,723 @@
+/// Trait for generic configuration information.
+pub trait ConfigItem {
+    /// Related errors.
+    type E;
```
Changed

studychao commented on src/dragonball/src/config_manager.rs:
```diff
@@ -0,0 +1,723 @@
+{
+    /// Constructor
+    pub fn new() -> Self {
+        ConfigInfos {
+            configs: Vec::new(),
```
```
pub struct ConfigInfos<T>
where
    T: ConfigItem + Clone,
{
    configs: Vec<T>,
}
```

Since `T` here doesn't have `Default trait`, we could not change this line to `ConfigInfos::default()`

studychao replied:
Yes, it works~~~I have changed this to `ConfigInfos::default()`

studychao commented on src/dragonball/src/config_manager.rs:
```diff
@@ -0,0 +1,723 @@
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct TokenBucketConfigInfo {
+    /// See TokenBucket::size.
+    pub size: u64,
```
Added

studychao commented on src/dragonball/src/config_manager.rs:
```diff
@@ -0,0 +1,723 @@
+
+impl TokenBucketConfigInfo {
+    fn resize(&mut self, n: u64) {
+        self.size /= n;
```
Check added

studychao commented:
Hi @jodh-intel , I have addressed all of the comments. Please take a look at them when you are free:)

jodh-intel commented on src/dragonball/src/device_manager/legacy.rs:
```diff
@@ -0,0 +1,148 @@
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Cannot add legacy device to Bus.
+    #[error("failure while managing legacy device")]
```
This is rather a generic description. How about something like:

```suggestion
    #[error("Bus error while managing legacy device")]
```

studychao replied:
Changed

jodh-intel reviewed (commented):
Thanks @studychao - a few more comments...

studychao commented:
Hi @jodh-intel, thanks for your review:) I'll address them this week.

studychao commented on src/dragonball/Makefile:
```diff
@@ -0,0 +1,28 @@
+default: build
+
+build:
+	cargo update -p vm-memory:0.8.0 --precise 0.7.0
```
There is a vm-memory dependency problem in our dependency libs and we are working hard to fix it.
We expect to fix this problem within 2 weeks but right now Dragonball part could not be compiled if vm-memory isn't fixed to 0.7.0.

So basically this is a temporary workaround here, and we will fix the problem and delete this line after we fix the dependency problem.

liubin replied:
@studychao It's better to include a FIXME/TODO here and fix it in the next coming PRs.

studychao replied:
@liubin @jodh-intel 
Thanks, a FIXME is added for this issue.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,851 @@
+/// Concrete GuestRegion type used by the VMM.
+pub type GuestRegionImpl = GuestRegionMmap;
+
+// Maximum number of working threads for memory pre-allocation.
```
I'll create an issue for this, so we could do the performance experiment when we could run Kata 3.0.

studychao replied:
#4382 added.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,851 @@
+// Control the actual number of pre-allocating threads. After several performance tests, we decide to use one thread to do pre-allocating for every 4G memory.
+const PRE_ALLOC_GRANULARITY: u64 = 32;
+
+// We don't have plan to support mainframe computer and only focus on PC servers.
+// 64 as max nodes should be enough for now.
+const MAX_NODE: u32 = 64;
```
Dragonball doesn't support s390x now. 

I'll create an issue for that and track the process for it.

studychao replied:
#4381 issue added.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,851 @@
+                    });
+
+                match handler {
+                    Err(e) => error!(
+                        "Failed to create working thread for async pre-allocation, {:?}",
+                        e
+                    ),
```
Yes, we design `prealloc` as a performance improvement function so we don't want it to influence the normal start-up process.

jodh-intel replied:
Ack. Maybe this should be a warning then that states that performance maybe be affected rather than making it a non-fatal error.

studychao replied:
warning added.

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,851 @@
+    }
+
+    // We set Transparent Huge Page (THP) through mmap to increase performance.
+    // In order to reduce the impact of page fault on performance, we start several threads (up to 16) to touch every 4k page of the memory region to manually do memory pre-allocation.
```
Thanks~ Changed~

studychao commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,851 @@
+#[derive(Debug, thiserror::Error)]
+pub enum AddressManagerError {
+    /// Invalid address space operation.
+    #[error("invalid address space operation")]
+    InvalidOperation,
+
+    /// Invalid address range.
+    #[error("invalid address space region (0x{0:x}, 0x{1:x})")]
+    InvalidAddressRange(u64, GuestUsize),
+
+    /// No resource available.
+    #[error("no available resource")]
```
Thanks, I have adjusted the code here to show the relation with AddressManager through error msg.

studychao commented on src/dragonball/src/device_manager/legacy.rs:
```diff
@@ -0,0 +1,148 @@
+            )));
+            let resources = [Resource::PioAddressRange {
+                base: port1,
+                size: 0x8,
```
comment added for that

liubin commented on src/dragonball/src/device_manager/vsock_dev_mgr.rs:
```diff
@@ -0,0 +1,285 @@
+const SUBSYSTEM: &str = "vsock_dev_mgr";
+// The flag of whether to use the shared irq.
+const USE_SHARED_IRQ: bool = true;
+// The flag of whether to use the shared irq.
```
```diff
+ // The flag of whether to use the generic irq.
- // The flag of whether to use the shared irq.
```

studychao replied:
Changed. Thanks:)

liubin commented on src/dragonball/src/device_manager/vsock_dev_mgr.rs:
```diff
@@ -0,0 +1,285 @@
+    #[error("the virtual machine instance ID is invalid")]
+    InvalidVMID,
+
+    /// The Context Identifier is already in use.
```
"Context Identifier" or "device ID"?

studychao replied:
Thanks, changed to `Device ID`.

studychao commented:
Hi @jodh-intel @liubin, I have addressed all your comments, plz take a look at them when you are free :)

Lots of thanks for your effort and suggestions.

liubin commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,865 @@
+}
+
+impl<'a> AddressSpaceMgrBuilder<'a> {
+    /// Create a new [`CreateParam`] object.
```
`CreateParam` -> `AddressSpaceMgrBuilder`?

studychao replied:
fixed

liubin commented on src/dragonball/src/device_manager/console_manager.rs:
```diff
@@ -0,0 +1,430 @@
+}
+
+/// Writer to process guest kernel dmesg.
+pub struct DmesgWriter {
```
Has `DmesgWriter` been used anywhere?

studychao replied:
Yes, it'll introduce in later PRs.

jodh-intel commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,865 @@
+/// Errors associated with virtual machine address space management.
+#[derive(Debug, thiserror::Error)]
+pub enum AddressManagerError {
+    /// Invalid address space operation.
+    #[error("invalid address space operation")]
```
Nit: I'd remove the comments as the error messages (should) make the comments entirely redundant.

studychao replied:
The error enum visibility here is pub and the variants in the pub enum should all have comments otherwise Rust will have a warning about it :(

jodh-intel commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,865 @@
+    InvalidAddressRange(u64, GuestUsize),
+
+    /// No address space available.
+    #[error("no available resource")]
```
If this error occurs, what's the full message that will be displayed? If it's just "no available resource", I think that's too generic: what resource is not available? The error message should specify more detail.

studychao replied:
I have split this error into two errors (`NoAvailableMemAddress` and `NoAvailableKvmSlot`) to be more specific about the lack of resources.

jodh-intel commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,865 @@
+        })
+    }
+
+    /// Enable/disable adding numbered suffix to memory file path.
```
It unclear from this comment why you might want to toggle this? What is this feature used for?

studychao replied:
Explanation added.

jodh-intel commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,865 @@
+        self.mem_suffix = enabled;
+    }
+
+    /// Enable/disable memory pre-allocation.
```
It might be helpful to specify why this is useful (performance?)

jodh-intel commented on src/dragonball/src/address_space_manager.rs:
```diff
@@ -0,0 +1,865 @@
+    }
+
+    fn configure_numa(&self, mmap_reg: &MmapRegion, node_id: u32) -> Result<()> {
+        let nodemask = (1 << node_id) as u64;
```
Again, this shift _could_ potentially lead to a crash, so safer to use `checked_shl()` (and add a test to assert that the method returns an error if node_id is too big.

jodh-intel commented on src/dragonball/src/device_manager/mod.rs:
```diff
@@ -0,0 +1,600 @@
+#[derive(Debug, thiserror::Error)]
+pub enum DeviceMgrError {
+    /// Invalid operation.
+    #[error("invalid operation")]
```
Again, if this is the message the user sees, can you add more detail ("invalid device manager operation" or similar?)

jodh-intel commented on src/dragonball/src/device_manager/mod.rs:
```diff
@@ -0,0 +1,600 @@
+    }
+
+    fn generate_kernel_boot_args(&mut self, kernel_config: &mut KernelConfigInfo) -> Result<()> {
+        assert!(!self.is_hotplug);
```
This looks potentially dangerous for production code. Could you return an error instead and add a unit test to assert this error path?

studychao replied:
Changed to return an error.
We'll introduce unit tests for the device manager part in later PRs so I'll add a unit test for this function in future PRs. #4410 issue is created for this.

jodh-intel commented on src/dragonball/src/device_manager/mod.rs:
```diff
@@ -0,0 +1,600 @@
+    /// Set the stream for guest kernel log.
+    ///
+    /// Note: com2 is used for guest kernel logging.
+    /// TODO: check whether it works with aarch64.
```
Maybe make this a GitHub issue URL as it's easy to forget TODO's and FIXME's in the code :smile:

@jongwu - fyi ^

jongwu replied:
I'm glad to see that Dragonball can support aarch64 and happy to test it on arm side, but for now I don't know how to start. Is there any guide to get me there? @studychao , thanks.

studychao replied:
Thanks, @jongwu :) 
Dragonball is still under development so it's hard to give a version for you to test it right now. We plan to submit all the code before the end of June, at that time you could have a try at it.
I'll let you know when it's ready for you to test and try~

xuejun-xj replied:
Thanks, @jongwu .

Arm Dragonball is part of rund, and rund is still under development, so it cannot be tested separately. We are glad to invite you to join us to improve its capabilities. For now, by adding unit tests, you may test some specific mechanisms on arm, like fdt, gic, its, vcpu and relevent procedures. We are open for any questions and feedbacks on arm Dragonball.

jongwu replied:
Sounds good. I can do some tests manually and it's better to add jenkins ci job for the runtime-rs in the future.

xuejun-xj replied:
Thanks, @jodh-intel :)

The GitHub issue URL for this TODO is #4409 .
We will complete this work in soon.

jodh-intel reviewed (commented):
Thanks @studychao - a few more comments...

ooi, are you planning to add tracing to dragonball?

studychao commented:
thanks @jodh-intel , I'll address them soon

studychao commented:
@jodh-intel Thanks for your review, I have addressed the latest comments.
Please take a look at them when you are free:)

studychao commented:
> ooi, are you planning to add tracing to dragonball?

For the tracing question, I am not sure what you mean by `tracing` here. Does that refer to `tracing` rust lib or the support for tracing tools like GDB?

If you mean GDB, we have achieved the GDB support for Dragonball but it will not be included in the first stage of Dragonball. It could be added in the future stages.

For `tracing` rust lib, we don't have support for it.

Hope that helps and please let me know if I misunderstood anything:)

studychao commented:
/test

jodh-intel reviewed (approved):
Thanks @studychao. One query regarding that default issue, but other than that...

lgtm

studychao commented:
@jodh-intel Thanks for your review:)
Sorry that I forgot to push the default issue-related code ... Now I have pushed the code and please take a look at it.
